rs/tests/nns/node_swap_test.rs

```diff
@@ -0,0 +1,211 @@
+    SystemTestGroup::new()
+        .with_timeout_per_test(OVERALL_TIMEOUT)
+        .with_setup(setup)
+        .add_test(ic_system_test_driver::driver::dsl::TestFunction::new(
+            "node_swaps",
+            test,
+        ))
+        .execute_from_args()
+}
+
+fn setup(env: TestEnv) {
+    // Ensure that the variable is set so that the test panics early
+    // if ic-admin isn't setup correctly.
+    fetch_ic_admin_path();
+
+    let caller = *TEST_USER1_PRINCIPAL;
+    let mut ic = InternetComputer::new()
+        .add_subnet(Subnet::new(ic_registry_subnet_type::SubnetType::System).add_nodes(3))
+        .with_node_operator(caller)
+        .with_node_provider(caller)
+        .with_unassigned_nodes(1);
+
+    ic.setup_and_start(&env)
+        .expect("Failed to start IC under test");
+
+    let snapshot = env.topology_snapshot();
+    let subnet = snapshot.root_subnet();
+
+    let customizations = NnsCustomizations {
+        registry_canister_init_payload: RegistryCanisterInitPayloadBuilder::new()
+            .enable_swapping_feature_globally()
+            .enable_swapping_feature_for_subnet(subnet.subnet_id)
+            .whitelist_swapping_feature_caller(caller)
+            .build(),
+        ..Default::default()
+    };
+
+    install_nns_with_customizations_and_check_progress(env.topology_snapshot(), customizations);
+}
+
+fn fetch_ic_admin_path() -> String {
+    std::env::var("IC_ADMIN_PATH").expect(
+        "IC admin isn't present in the environment variables and is required for this test to run",
+    )
+}
+
+fn test(env: TestEnv) {
+    block_on(test_inner(env))
+}
+
+async fn test_inner(env: TestEnv) {
+    let snapshot = env.topology_snapshot();
+
+    let unassigned_node = snapshot.unassigned_nodes().next().unwrap();
+    let subnet = snapshot.subnets().next().unwrap();
+    let mut nodes_iter = subnet.nodes();
+
+    let assigned_node = nodes_iter.next().unwrap();
+
+    let next_nns_node = nodes_iter.next().unwrap();
+    let url = format!("http://[{}]:8080", next_nns_node.get_ip_addr());
+    ic_admin_swap_nodes(
+        &url,
+        &TEST_USER1_KEYPAIR.to_pem(),
+        assigned_node.node_id,
+        unassigned_node.node_id,
+        &env,
+    )
+    .await
+    .unwrap();
+
+    snapshot.block_for_newer_registry_version().await.unwrap();
+
+    let new_unassigned_node = snapshot.unassigned_nodes().next().unwrap();
+    // Expect the new unassigned node to be the previously assigned one
+    // which should happen if the canister didn't return any errors.
+    assert_eq!(new_unassigned_node.node_id, assigned_node.node_id);
+
+    // Expect the previously unassigned node to be a member of
+    // a subnet it was directly swapped into.
+    let new_node_in_subnet = snapshot
+        .subnets()
+        .next()
+        .unwrap()
+        .nodes()
+        .find(|node| node.node_id == unassigned_node.node_id)
+        .expect("Expected the unassigned node to become a member of a subnet");
```

Comment:
Can you try to add some breadcrumbs? Don't just explode: leave behind a black box.